backend/lib/components/OAuth2/Facebook.js

```diff
@@ -0,0 +1,26 @@
+const { Social } = require("./Social");
+
+// reference: https://developers.facebook.com/docs/facebook-login/manually-build-a-login-flow/
+const authUrlBase = "https://www.facebook.com/v9.0/dialog/oauth";
```

Comment:
@mannykary when I define static properties in the children classes, my parent class member function cannot access them, (e.g. static authUrlBase in Facebook and buildOauthUrl() from Social), so I use `this.authUrlBase` currently so Social can access it (while this way doesn't guarantee children classes define these variables, any ideas how could I make it better?

Reply:
Ah that's a good point. Unfortunately ES6 doesn't support interfaces, so we can't really enforce subclasses not implementing these variables. I suppose you could just pass the variables into the super constructor from the child class and then do a check in the super constructor to make sure all arguments passed in are defined, and then set them as instance variables on the superclass. Not that pretty, but it would work.

```
class BaseClass {
  constructor(a, b, c) {
    // Check all args are defined
    if (!a || !b || !c) throw new TypeError('Must define all arguments');
    this.a = a;
    this.b = b;
    this.c = c;
  }
}

class ChildClass extends BaseClass {
  constructor(a) {
    super(a, 'def', 'hij');
  }
}

> const c = new ChildClass(). // pass no arguments - should throw error
Uncaught TypeError: Must define all arguments
    at new BaseClass (repl:4:31)
    at new ChildClass (repl:3:5)
> const c1 = new ChildClass('abc')
undefined
> c1
ChildClass { a: 'abc', b: 'def', c: 'hij' }

```